\documentclass{article}
<<style, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@

\title{Copy Number approach - seqCNA}
\date{Modified :26 March, 2014. Compiled: \today}

\begin{document}
\maketitle

\section{method}
Here we will review the Bioconductor package seqCNA

David Mosen-Ansorena, N. T., Silvia Veganzones,VirginiaDelaOrden, 
Maria Luisa Maestro and a. AnaMAransay (2014). "seqCNA: an R package 
for DNA copy number analysis in cancer using high-throughput sequencing."
 BMCGenomics.  


\section{code}

<<package-load,message=FALSE>>=
library(seqCNA)
datafiles <- file.path("/shared","silo_researcher","Morgan_M","BioC","sonali",
                      "analysis","inst","extdata")
@
Loading test data and running seqCNA on test data.
This tumor sample and normal sample contain data only from "chr4".
<<Paul-data-load>>=
tumor <- file.path(datafiles, "tumorA.chr4.bam")
normal <- file.path(datafiles, "normalA.chr4.bam")
##runSeqsumm is run only once.
#runSeqsumm(summ.win=10, 
#           file=tumor, 
#           output.file="tumorA_chr4_seqsumm_out.txt",
#           samtools.path="samtools")

# runSeqsumm(summ.win=10, 
#            file=normal, 
#            output.file="normalA_chr4_seqsumm_out.txt",
#            samtools.path="samtools")

##note: by default it has written the two files in the same folder
## as the bam files!!

tumordata<- read.table(file.path(datafiles,"tumorA_chr4_seqsumm_out.txt"), 
                       sep="\t", header=TRUE)
normaldata<- read.table(file.path(datafiles,"normalA_chr4_seqsumm_out.txt"), 
                       sep="\t", header=TRUE)
tumordata<- tumordata[which(tumordata$chrom=="chr4"),]
normaldata<- normaldata[which(normaldata$chrom=="chr4"),]
head(tumordata)
@

When we run \Rfunction{readSeqsumm} it generates the following lines in the output
file for other chromosomes.Thus we have subsetted to look only at chr4.
<<noentry,eval=FALSE>>=
#chr5    0       0       0       0
#chr6    0       0       0       0
@

<<Realdata, eval=FALSE,echo=FALSE>>=
tumor <- file.path(datafiles, "tumor_srx036691.bam")
normal <- file.path(datafiles, "normal_srx113635.bam")

##runSeqsumm is run only once. 
#runSeqsumm(summ.win=10, 
#           file=tumor, 
#           output.file="tumor_seqsumm_out.txt",
#           samtools.path="samtools")
 
# runSeqsumm(summ.win=10, 
#            file=normal, 
#            output.file="normal_seqsumm_out.txt",
# samtools.path="samtools")
 
##note: by default it has written the two files in the same folder 
## as the bam files!!
 
tumordata<- read.table(file.path(datafiles,"tumor_seqsumm_out.txt"), 
                       sep="\t", header=TRUE)
normaldata<- read.table(file.path(datafiles,"normal_seqsumm_out.txt"), 
                       sep="\t", header=TRUE)
@
Running seqCNA according to manual.
<<seqCNA>>= 
##read in the summarized data
rco = readSeqsumm(build="hg19",
                    tumour.data=tumordata,
                    normal.data=normaldata)
 
##apply the trimming and mapping quality filters                  
rco =applyFilters(rco, trim.filter=1, mapq.filter=2) 
rco= runSeqnorm(rco)
rco=runGLAD(rco)

plotCNProfile(rco)
 
rco = applyThresholds(rco, seq(-0.8,4,by=0.8), 1)

plotCNProfile(rco)

summary(rco)

#writeCNProfile(rco, dir)
@

\end{document}
